Statistical validation of peptide assignments for MS/MS Proteomics data


## Version

PeptidProphet v5.01


## Usage

`philosopher peptideprophet [flags] [files]`


## Flags

`--accmass`

Use Accurate Mass model binning.

`--clevel`

Set Conservative Level in neg_stdev from the neg_mean, low numbers are less conservative, high numbers are more conservative.

`--combine`

Combine the results from PeptideProphet into a single result file.

`--database`

Path to the database file.

`--decoy`

Semi-supervised mode, protein name prefix to identify Decoy entries.

`--decoyprobs`

Compute possible non-zero probabilities for Decoy entries on the last iteration.

`--enzyme`

Enzyme used in sample

`--exclude`

Exclude deltaCn*, Mascot*, and Comet* results from results (default Penalize * results).

`--expectscore`

Use expectation value as the only contributor to the f-value for modeling.

`--forcedistr`

Bypass quality control checks, report model despite bad modeling.

`--glyc`

Enable peptide Glyco motif model.

`--icat`

Apply ICAT model (default Auto detect ICAT).

`--ignorechg`

Can be used multiple times to specify all charge states to exclude from modeling.

`--instrwarn`

Warn and continue if combined data was generated by different instrument models.

`--leave`

Leave alone deltaCn*, Mascot*, and Comet* results from results (default Penalize * results).

`--maldi`

Enable MALDI mode.

`--masswidth`

model mass width (default 5.0)

`--minpintt`

Minimum number of NTT in a peptide used for positive pI model (default 2).

`--minpiprob`

Minimum probability after first pass of a peptide used for positive pI model (default 0.9).

`--minprob`

Report results with minimum probability <number> (default 0.05).

`--minrtntt`

Minimum number of NTT in a peptide used for positive RT model.

`--minrtprob`

Minimum probability after first pass of a peptide used for positive RT model.

`--neggamma`

Use Gamma distribution to model the negative hits.

`--noicat`

Do no apply ICAT model (default Auto detect ICAT).

`--nomass`

Disable mass model.

`--nonmc`

Disable NMC missed cleavage model.

`--nonparam`

Use semi-parametric modeling, must be used in conjunction with --decoy option.

`--nontt`

Disable NTT enzymatic termini model.

`--optimizefval`

(SpectraST only) optimize f-value function f(dot,delta) using PCA.

`--output`

Output name prefix (default "interact").

`--perfectlib`

Multiply by SpectraST library probability.

`--phospho`

Enable peptide Phospho motif model.

`--pi`

Enable peptide pI model.

`--ppm`

Use ppm mass error instead of Dalton for mass modeling.

`--rt`

Enable peptide RT model.

`--rtcat`

Enable peptide RT model, use <rtcatalog_file> peptide RTs when available as the theoretical value.

`--zero`

Report results with minimum probability 0 (default 0.05).


## Example

Process two pepXML files, combining them into a single output called _combined_samples_. The analysis will compute possible non-zero probabilities for decoy entries using an accurate mass model for binning.

`philosopher peptideprophet --database db.fasta --combine --decoyprobs  --accmass --nonparam --output combined_samples sample1.pepxml sample2.pepxml`


## FAQ

_Do I need TPP installed for running this ?_

No
